### PR TITLE
Prompt Scheduling for Diffusers

### PR DESCRIPTION
**Pull request by AI-Casanova**
[prompt:scheduling;0.5] syntax from A1111 implemented for diffusers callback.
xyz_grid-xyz_grid-0001-juggernautXL_version5-A photograph of a man woman 0 8k

Caching is enabled for fewer calls through the Text Encoders.

Batching logic is also changed for fewer calls, even without scheduling.

Tested on SDXL and SD1.5, other Diffusers models appear to have the relevant callback options so no collisions expected.

### CODE REVIEW

File: modules/processing_diffusers.py [hunk 1]
```diff
@@ -293,36 +302,29 @@ def set_pipeline_args(model, prompts: list, negative_prompts: list, prompts_2: t
...
         if 'prompt' in possible:
-            if hasattr(model, 'text_encoder') and 'prompt_embeds' in possible and prompt_embed is not None:
-                if type(pooled) == list:
-                    pooled = pooled[0]
-                if type(negative_pooled) == list:
-                    negative_pooled = negative_pooled[0]
-                args['prompt_embeds'] = prompt_embed
+            if hasattr(model, 'text_encoder') and 'prompt_embeds' in possible and p.prompt_embeds[0] is not None:
+                args['prompt_embeds'] = p.prompt_embeds[0]
                 if 'XL' in model.__class__.__name__:
```
Comment (vladmandic) [inline modules/processing_diffusers.py:319]: this is becoming an issue as there are other pipelines that define classes that inherit from base class, so class name becomes meaningless. this works for base SDXL, but not for any inherited classes. can we make this check a bit smarter?

File: modules/prompt_parser_diffusers.py [hunk 2]
```diff
@@ -58,32 +58,48 @@ def expand_textual_inversion_token_ids_if_necessary(self, token_ids: typing.List
...
-    if positive_pooleds is not None and shared.sd_model_type == "sdxl":
-        positive_pooleds = torch.cat(positive_pooleds, dim=0)
-    if negative_pooleds is not None and shared.sd_model_type == "sdxl":
-        negative_pooleds = torch.cat(negative_pooleds, dim=0)
-    return prompt_embeds, positive_pooleds, negative_embeds, negative_pooleds
+        positive_schedule, scheduled = get_prompt_schedule(p, prompts[0], steps)
+        negative_schedule, neg_scheduled = get_prompt_schedule(p, negative_prompts[0], steps)
+        p.scheduled_prompt = scheduled or neg_scheduled
+
+        p.prompt_embeds = []
```
Comment (vladmandic) [inline modules/prompt_parser_diffusers.py:79]: ah, no local vars anymore, you're setting in on processing object directly - i'm ok with that.

File: modules/processing_diffusers.py [hunk 3]
```diff
@@ -89,6 +89,15 @@ def diffusers_callback(_pipe, step: int, timestep: int, kwargs: dict):
         if kwargs.get('latents', None) is None:
             return kwargs
         kwargs = correction_callback(p, timestep, kwargs)
+        if p.scheduled_prompt:
+            try:
+                i = (step + 1) % len(p.prompt_embeds)
+                kwargs["prompt_embeds"] = p.prompt_embeds[i][0:1].repeat(1, kwargs["prompt_embeds"].shape[0], 1).view(
```
Comment (vladmandic) [inline modules/processing_diffusers.py:95]: can we make this conditional if all prompts are actually equal? typically they are, but there are scenarios where different prompts may be desirable, for example getting into video stuff where different prompts are allowed.

File: modules/prompt_parser_diffusers.py [hunk 4]
```diff
@@ -58,32 +58,48 @@ def expand_textual_inversion_token_ids_if_necessary(self, token_ids: typing.List
...
+    temp = []
+    schedule = prompt_parser.get_learned_conditioning_prompt_schedules([prompt], steps)[0]
+    for chunk in schedule:
+        for s in range(steps):
+            if len(temp) < s + 1 <= chunk[0]:
+                temp.append(chunk[1])
+    return temp, len(schedule) > 1
+
+def encode_prompts(pipe, p,  prompts: list, negative_prompts: list, steps: int, step: int = 1, clip_skip: typing.Optional[int] = None):
     if 'StableDiffusion' not in pipe.__class__.__name__ and 'DemoFusion':
```
Comment (vladmandic) [inline modules/prompt_parser_diffusers.py:71]: same comment as before, you can see how its becoming an issue, can't list ever class name that ever comes up.

File: modules/prompt_parser_diffusers.py [hunk 5]
```diff
@@ -58,32 +58,48 @@ def expand_textual_inversion_token_ids_if_necessary(self, token_ids: typing.List
...
-    if negative_embeds is not None:
-        negative_embeds = torch.cat(negative_embeds, dim=0)
-    if positive_pooleds is not None and shared.sd_model_type == "sdxl":
-        positive_pooleds = torch.cat(positive_pooleds, dim=0)
-    if negative_pooleds is not None and shared.sd_model_type == "sdxl":
-        negative_pooleds = torch.cat(negative_pooleds, dim=0)
-    return prompt_embeds, positive_pooleds, negative_embeds, negative_pooleds
+        positive_schedule, scheduled = get_prompt_schedule(p, prompts[0], steps)
+        negative_schedule, neg_scheduled = get_prompt_schedule(p, negative_prompts[0], steps)
+        p.scheduled_prompt = scheduled or neg_scheduled
```
Comment (vladmandic) [inline modules/prompt_parser_diffusers.py:77]: maybe some debug output using existing `SD_PROMPT_DEBUG`?

File: modules/processing_diffusers.py (hunk 3 above)
Comment (AI-Casanova) [reply modules/processing_diffusers.py:95]: I can dig into it. For some reason there's an internal `torch.repeat` in diffusers, that returns size `(2, 77, 2048)` on a batch size 1 with input of size `(1, 77, 2048)` but seemingly only on the positive prompt (but I was using blank negative)

I'll see what I can find out about how multi prompt batches would shake out.

File: modules/prompt_parser_diffusers.py (hunk 2 above)
Comment (AI-Casanova) [reply modules/prompt_parser_diffusers.py:79]: I was having trouble conceptualizing a better passthrough to the callback function.

**Pull request merged**